// @azure/web-pubsub/dist-esm/src/generated/operations/healthApi.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/** Class containing HealthApi operations. */
export class HealthApiImpl {
    /**
     * Initialize a new instance of the class HealthApi class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Get service health status.
     * @param options The options parameters.
     */
    getServiceStatus(options) {
        return this.client.sendOperationRequest({ options }, getServiceStatusOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getServiceStatusOperationSpec = {
    path: "/api/health",
    httpMethod: "HEAD",
    responses: { 200: {}, default: {} },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.endpoint],
    serializer
};
