// @azure/web-pubsub-client/dist-esm/src/websocket/websocketClient.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import WebSocket from "ws";
export class WebSocketClient {
    constructor(uri, protocolName) {
        this._socket = new WebSocket(uri, protocolName);
        this._socket.binaryType = "arraybuffer";
    }
    onopen(fn) {
        this._socket.onopen = fn;
    }
    onclose(fn) {
        this._socket.onclose = (event) => fn(event.code, event.reason);
    }
    onerror(fn) {
        this._socket.onerror = (event) => fn(event.error);
    }
    onmessage(fn) {
        this._socket.onmessage = (event) => fn(event.data);
    }
    close(code, reason) {
        this._socket.close(code, reason);
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    send(data, _) {
        return new Promise((resolve, reject) => {
            this._socket.send(data, (err) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    isOpen() {
        return this._socket.readyState === WebSocket.OPEN;
    }
}
export class WebSocketClientFactory {
    create(uri, protocolName) {
        return new WebSocketClient(uri, protocolName);
    }
}
