// @azure/web-pubsub/dist-esm/src/generated/models/mappers.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
export const ErrorDetail = {
    type: {
        name: "Composite",
        className: "ErrorDetail",
        modelProperties: {
            code: {
                serializedName: "code",
                type: {
                    name: "String"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            },
            target: {
                serializedName: "target",
                type: {
                    name: "String"
                }
            },
            details: {
                serializedName: "details",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ErrorDetail"
                        }
                    }
                }
            },
            inner: {
                serializedName: "inner",
                type: {
                    name: "Composite",
                    className: "InnerError"
                }
            }
        }
    }
};
export const InnerError = {
    type: {
        name: "Composite",
        className: "InnerError",
        modelProperties: {
            code: {
                serializedName: "code",
                type: {
                    name: "String"
                }
            },
            inner: {
                serializedName: "inner",
                type: {
                    name: "Composite",
                    className: "InnerError"
                }
            }
        }
    }
};
export const ClientTokenResponse = {
    type: {
        name: "Composite",
        className: "ClientTokenResponse",
        modelProperties: {
            token: {
                serializedName: "token",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubCloseAllConnectionsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubCloseAllConnectionsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubGenerateClientTokenExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubGenerateClientTokenExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubSendToAllExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubSendToAllExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubCloseConnectionExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubCloseConnectionExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubConnectionExistsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubConnectionExistsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubSendToConnectionExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubSendToConnectionExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubRemoveConnectionFromAllGroupsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubRemoveConnectionFromAllGroupsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubGroupExistsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubGroupExistsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubCloseGroupConnectionsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubCloseGroupConnectionsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubSendToGroupExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubSendToGroupExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubRemoveConnectionFromGroupExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubRemoveConnectionFromGroupExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubAddConnectionToGroupExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubAddConnectionToGroupExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubRevokePermissionExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubRevokePermissionExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubCheckPermissionExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubCheckPermissionExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubGrantPermissionExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubGrantPermissionExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubUserExistsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubUserExistsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubCloseUserConnectionsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubCloseUserConnectionsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubSendToUserExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubSendToUserExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubRemoveUserFromAllGroupsExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubRemoveUserFromAllGroupsExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubRemoveUserFromGroupExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubRemoveUserFromGroupExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export const WebPubSubAddUserToGroupExceptionHeaders = {
    type: {
        name: "Composite",
        className: "WebPubSubAddUserToGroupExceptionHeaders",
        modelProperties: {
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
