// @azure/web-pubsub-socket.io/dist/common/rest-service-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestServiceClient = void 0;
const web_pubsub_1 = require("@azure/web-pubsub");
const operation_spec_1 = require("./azure-api/operation-spec");
const utils_1 = require("./utils");
const core_auth_1 = require("@azure/core-auth");
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const core_tracing_1 = require("@azure/core-tracing");
/** @internal */
const tracingClient = (0, core_tracing_1.createTracingClient)({
    namespace: "Microsoft.WebPubSub",
    packageName: "@azure/web-pubsub-socket.io",
});
const debug = (0, utils_1.debugModule)("wps-sio-ext:common:rest-service-client");
/**
 * Communicate with Azure Web PubSub service via traditional REST API.
 */
class RestServiceClient extends web_pubsub_1.WebPubSubServiceClient {
    async invoke(message, bodyHandler, options) {
        const onResponse = (rawResponse, flatResponse, error) => {
            if (error || rawResponse.status !== 200) {
                // Log and do nothing, let it timeout
                debug(`broadcastWithAck response status code = ${rawResponse["status"]}, error = ${error},\
rawResponse = ${JSON.stringify(rawResponse)}`);
                return;
            }
            if (rawResponse.browserStreamBody) {
                // Browser stream
                const reader = rawResponse.browserStreamBody["getReader"]();
                reader.read().then(function processText({ done, value }) {
                    bodyHandler(value, done);
                });
            }
            else {
                const stream = rawResponse["readableStreamBody"];
                stream.on("end", () => {
                    bodyHandler(undefined, true);
                });
                stream.on("data", (chunk) => {
                    bodyHandler(chunk.toString(), false);
                });
            }
        };
        const operationArguments = {
            ...options,
            hub: this.hubName,
            message: message,
            options: { onResponse: onResponse },
            filter: options?.filter,
            contentType: options?.contentType,
        };
        await this["client"].sendOperationRequest(operationArguments, (0, operation_spec_1.getInvokeOperationSpec)(this.endpoint));
    }
    /**
     * Generate a token for a client to connect to the Azure Web PubSub service.
     *
     * @param options - Additional options
     */
    async getClientAccessToken(options = {}) {
        return tracingClient.withSpan("getClientAccessToken", options, async (updatedOptions) => {
            const endpoint = this.endpoint.endsWith("/") ? this.endpoint : this.endpoint + "/";
            const baseUrl = `${endpoint}clients/socketio/hubs/${this.hubName}`;
            const credential = this["credential"];
            const innerClient = this["client"];
            let token;
            if ((0, core_auth_1.isTokenCredential)(credential)) {
                const response = await innerClient.webPubSub.generateClientToken(this.hubName, updatedOptions);
                token = response.token;
                if (token === null || token === undefined) {
                    throw new Error(`Response from "generateClientToken" doesn't contain valid property "token"`);
                }
            }
            else {
                const key = credential.key;
                const payload = {
                    role: options?.roles,
                    "webpubsub.group": options?.groups,
                    userId: options?.userId,
                };
                const signOptions = {
                    audience: baseUrl,
                    expiresIn: options?.expirationTimeInMinutes === undefined ? "1h" : `${options.expirationTimeInMinutes}m`,
                    algorithm: "HS256",
                };
                if (options?.userId) {
                    signOptions.subject = options?.userId;
                }
                token = jsonwebtoken_1.default.sign(payload, key, signOptions);
            }
            return {
                token,
                baseUrl,
                url: `${baseUrl}?access_token=${token}`,
            };
        });
    }
}
exports.RestServiceClient = RestServiceClient;
