// @azure/web-pubsub-socket.io/dist/common/utils.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.debugModule = exports.getSioMiddlewareFromExpress = exports.writeResponse = exports.toOptionsString = exports.toString = exports.toAsync = exports.getWebPubSubServiceClient = exports.getWebPubSubServiceCaller = exports.addProperty = exports.debug = exports.T = void 0;
const debug_1 = __importDefault(require("debug"));
exports.debugModule = debug_1.default;
const rest_service_client_1 = require("./rest-service-client");
const serverProxies_1 = require("../serverProxies");
const T = (now) => `${now.toLocaleString().replace(" AM", "").replace(" PM", "")}:${now.getMilliseconds().toString().padStart(3, '0')}`; // prettier-ignore
exports.T = T;
debug_1.default.log = (msg, ...args) => {
    const timestamp = (0, exports.T)(new Date());
    console.log(`[${timestamp}] ${msg}`, ...args);
};
exports.debug = (0, debug_1.default)("wps-sio-ext:common:utils");
function addProperty(o, p, f) {
    Object.defineProperty(o, p, {
        value: f,
        writable: true,
        enumerable: false,
        configurable: true,
    });
}
exports.addProperty = addProperty;
function checkRequiredKeys(options, requiredKeys) {
    for (const key of requiredKeys) {
        if (options[key] === undefined || options[key] === null || options[key] === "")
            return false;
    }
    return true;
}
function getWebPubSubServiceCaller(options) {
    (0, exports.debug)(`getWebPubSubServiceCaller, ${JSON.stringify(options)}`);
    // if owns connection string, handle as `AzureSocketIOOptions`
    if (Object.keys(options).indexOf("connectionString") !== -1) {
        (0, exports.debug)(`getWebPubSubServiceCaller, use connection string`);
        const requiredKeys = ["connectionString", "hub"];
        if (checkRequiredKeys(options, requiredKeys)) {
            const args = [options["connectionString"], options.hub];
            return serverProxies_1.InprocessServerProxy.fromConnectionString(...args);
        }
        throw new Error(`Expect valid options with keys ${requiredKeys} are expected, but got null or empty value`);
    }
    else {
        (0, exports.debug)(`getWebPubSubServiceCaller, use credential`);
        const requiredKeys = ["endpoint", "credential", "hub"];
        if (checkRequiredKeys(options, requiredKeys)) {
            const args = [options["endpoint"], options["credential"], options.hub];
            return new serverProxies_1.InprocessServerProxy(...args);
        }
        throw new Error(`Expect valid options with keys ${requiredKeys} are expected, but got null or empty value`);
    }
}
exports.getWebPubSubServiceCaller = getWebPubSubServiceCaller;
function getWebPubSubServiceClient(options) {
    (0, exports.debug)(`getWebPubSubServiceClient, ${JSON.stringify(options)}`);
    // if owns connection string, handle as `AzureSocketIOOptions`
    if (Object.keys(options).indexOf("connectionString") !== -1) {
        (0, exports.debug)(`getWebPubSubServiceClient, use connection string`);
        const requiredKeys = ["connectionString", "hub"];
        if (checkRequiredKeys(options, requiredKeys)) {
            const args = [options["connectionString"], options.hub];
            return new rest_service_client_1.RestServiceClient(...args, { reverseProxyEndpoint: options.reverseProxyEndpoint });
        }
        throw new Error(`Expect valid options with keys ${requiredKeys} are expected, but got null or empty value`);
    }
    else {
        (0, exports.debug)(`WebPubSubServiceClient, use credential`);
        const requiredKeys = ["endpoint", "credential", "hub"];
        if (checkRequiredKeys(options, requiredKeys)) {
            const args = [options["endpoint"], options["credential"], options.hub];
            return new rest_service_client_1.RestServiceClient(...args, { reverseProxyEndpoint: options.reverseProxyEndpoint });
        }
        throw new Error(`Expect valid options with keys ${requiredKeys} are expected, but got null or empty value`);
    }
}
exports.getWebPubSubServiceClient = getWebPubSubServiceClient;
/**
 * Convert a sync function with callback parameter to its async form.
 * @param syncFunc - a sync function with callback as its last parameter
 * @returns the async function converted from sync function `syncFunc`
 */
function toAsync(syncFunc) {
    return (...args) => new Promise((resolve, reject) => {
        try {
            syncFunc(...args, (ret) => {
                resolve(ret);
            });
        }
        catch (error) {
            reject(error);
        }
    });
}
exports.toAsync = toAsync;
/**
 * Stringify a set or list of string .
 * @param set - a set or list of string. Example: Set\<string\>{"a", "b"}
 * @returns the stringified set. Example: "{ "a", "b" }"
 */
function toString(set) {
    // if (set.rooms) return `${{toString(set.rooms)}}`
    return set ? `{ "${[...set].join('", "')}" }` : "{}";
}
exports.toString = toString;
// `JSON.stringify` cannot stringify `BroadcastOptions` completely. `rooms` and `except` detail will be lost.
function toOptionsString(option) {
    return `{rooms: ${toString(option.rooms)}, except: ${toString(option.except)},\
flags: ${JSON.stringify(option.flags)}}`;
}
exports.toOptionsString = toOptionsString;
function writeResponse(res, statusCode, message, contentType = "application/json") {
    res.writeHead(statusCode, { "Content-Type": contentType });
    res.end(JSON.stringify(message));
}
exports.writeResponse = writeResponse;
function getSioMiddlewareFromExpress(middleware) {
    return (socket, next) => {
        return middleware(socket.request, {}, next);
    };
}
exports.getSioMiddlewareFromExpress = getSioMiddlewareFromExpress;
