// @azure/web-pubsub-socket.io/dist/EIO/components/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TUNNEL_PATH = exports.EIO_CONNECTION_ERROR = exports.CONNECTION_ERROR_WEBPUBSUB_MESSAGE = exports.CONNECTION_ERROR_WEBPUBSUB_CODE = exports.CONNECTION_ERROR_EVENT_NAME = exports.WEBPUBSUB_CONNECT_RESPONSE_FIELD_NAME = exports.WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME = exports.WEBPUBSUB_TRANSPORT_NAME = void 0;
exports.WEBPUBSUB_TRANSPORT_NAME = "webpubsub";
exports.WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME = "webPubSubClientConnection";
exports.WEBPUBSUB_CONNECT_RESPONSE_FIELD_NAME = "socketio";
// Reference: https://github.com/socketio/engine.io/blob/123b68c04f9e971f59b526e0f967a488ee6b0116/README.md?plain=1#L219
exports.CONNECTION_ERROR_EVENT_NAME = "connection_error";
exports.CONNECTION_ERROR_WEBPUBSUB_CODE = 6;
exports.CONNECTION_ERROR_WEBPUBSUB_MESSAGE = "Web PubSub Extension Internal Error";
exports.EIO_CONNECTION_ERROR = {
    UNKNOWN_TRANSPORT: 0,
    UNKNOWN_SID: 1,
    BAD_HANDSHAKE_METHOD: 2,
    BAD_REQUEST: 3,
    FORBIDDEN: 4,
    UNSUPPORTED_PROTOCOL_VERSION: 5,
};
exports.TUNNEL_PATH = "/eventhandler/";
